### PR TITLE
adds support to initiate streams upload using tus

### PR DESCRIPTION
**Pull request by bbbshah**
Adds support to streams to initiate tus video upload.

## Description

The current api client doesn't support streams: initiate tus video upload.
This PR adds a new func `StreamInitiateTUSVideoUpload` to `streams.go`.

## Has your change been tested?

Added unit tests for the newly added functions.

## Types of changes

What sort of change does your code introduce/modify?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] This change is using publicly documented in [cloudflare/api-schemas](https://github.com/cloudflare/api-schemas) 
      and relies on stable APIs.

[1]: https://help.github.com/articles/closing-issues-using-keywords/

### CODE REVIEW

**Comment by bbbshah** [comment]
@jacobbednarz Would you be able to take a look at this PR when you get a chance? Thank you in advance!

File: stream.go [hunk 1]
```diff
@@ -171,6 +190,55 @@ type StreamSignedURLParameters struct {
 	AccessRules  []StreamAccessRule `json:"accessRules,omitempty"`
 }
 
+type StreamInitiateTUSUploadParameters struct {
+	DirectUserUpload bool
+	TusResumable     TusProtocolVersion
+	UploadLength     int64
+	UploadCreator    string
+	Metadata         TUSUploadMetadata
+}
```
Comment (jacobbednarz) [inline stream.go]:
```suggestion
type StreamInitiateTUSUploadParameters struct {
	DirectUserUpload bool               `url:"direct_user,omitempty"`
	TusResumable     TusProtocolVersion `url:"-"`
	UploadLength     int64              `url:"-"`
	UploadCreator    string             `url:"-"`
	Metadata         TUSUploadMetadata  `url:"-"`
}
```

File: stream.go [hunk 2]
```diff
@@ -328,7 +396,56 @@ func (api *API) StreamListVideos(ctx context.Context, params StreamListParameter
...
+		return StreamInitiateTUSUploadResponse{}, ErrMarshallingTUSMetadata
+	}
+	if metadataTusCsv != "" {
+		headers.Set("Upload-Metadata", metadataTusCsv)
+	}
+
+	uri := fmt.Sprintf("/accounts/%s/stream", rc.Identifier)
+	if params.DirectUserUpload {
+		uri += "?direct_user=true"
+	}
```
Comment (jacobbednarz) [inline stream.go]:
```suggestion
	uri := buildURI(fmt.Sprintf("/accounts/%s/stream", rc.Identifier), params)
```

File: stream_test.go [hunk 3]
```diff
@@ -524,3 +524,98 @@ func TestStream_CreateSignedURL(t *testing.T) {
...
+		// Make sure Tus-Resumable header is set
+		assert.Equal(t, "1.0.0", r.Header.Get("Tus-Resumable"))
+		// Make sure Upload-Length header is set
+		assert.Equal(t, "123", r.Header.Get("Upload-Length"))
+		// set the response headers
+		// if query param direct_user=true, then return the direct url in the header
+		if r.URL.Query().Get("direct_user") == "true" {
+			w.Header().Set("Location", "https://upload.videodelivery.net/tus/90c68cb5cd4fd5350b1962279c90bec0?tusv2=true")
+		} else {
+			w.Header().Set("Location", "https://production.gateway.api.cloudflare.com/client/v4/accounts/test-account-id/media/278f2a7e763c73dedc064b965d2cfbed?tusv2=true")
```
Comment (jacobbednarz) [inline stream_test.go]:
```suggestion
			w.Header().Set("Location", "https://api.cloudflare.com/client/v4/accounts/"+testAccountID+"/media/278f2a7e763c73dedc064b965d2cfbed?tusv2=true")
```

File: stream_test.go [hunk 4]
```diff
@@ -524,3 +524,98 @@ func TestStream_CreateSignedURL(t *testing.T) {
...
+	// Make sure Upload-Length header is set
+	_, err = client.StreamInitiateTUSVideoUpload(context.Background(), AccountIdentifier(testAccountID), params)
+	if assert.Error(t, err) {
+		assert.Equal(t, ErrMissingUploadLength, err)
+	}
+	params.UploadLength = 123
+
+	out, err := client.StreamInitiateTUSVideoUpload(context.Background(), AccountIdentifier(testAccountID), params)
+	if assert.NoError(t, err) {
+		assert.Equal(t, "https://production.gateway.api.cloudflare.com/client/v4/accounts/test-account-id/media/278f2a7e763c73dedc064b965d2cfbed?tusv2=true", out.ResponseHeaders.Get("Location"))
```
Comment (jacobbednarz) [inline stream_test.go]:
```suggestion
		assert.Equal(t, "https://api.cloudflare.com/client/v4/accounts/"+testAccountID+"/media/278f2a7e763c73dedc064b965d2cfbed?tusv2=true", out.ResponseHeaders.Get("Location"))
```

File: stream.go (hunk 1 above)
Comment (jacobbednarz) [reply stream.go]: this allows us to use the struct for query parameter building in `buildURI` instead of manually performing it.

**Comment by jacobbednarz** [comment]
this looks good, thanks!

for future PRs, i recommend clicking the "allow maintainers to make changes" checkbox to save the back and forth here and we can apply small edits like the above ones without having to do the extra round trip.

**Pull request merged**